Basic implementation of STscript commands

bolmin70 opened the pull request:
Added two STscript commands to enable and disable sorcery scripts. 

Added a script ID number to each sorcery script in the tab, to be used in the commands.

bolmin70 committed:
Basic implementation of STscript commands for enabling and disabling sorcery scripts.

Added two STscript commands to enable and disable sorcery scripts. Added a script ID number to each sorcery script in the tab, to be used as handles in the commands.

p-e-w commented on main.js:
```diff
@@ -10,13 +10,17 @@ import { Handlebars, hljs } from "../../../../lib.js";
 import { NAME } from "./common.js";
 import { loadSettings } from "./settings.js";
 
+import { extension_settings } from "../../../extensions.js";
```
Unused

p-e-w commented on main.js:
```diff
@@ -10,13 +10,17 @@ import { Handlebars, hljs } from "../../../../lib.js";
 
 const settings = loadSettings();
 
+
```
:fire:

p-e-w commented on main.js:
```diff
@@ -201,6 +205,75 @@ function clickHandlerHack() {
 hljs.registerLanguage("sorcery-stscript", () => hljs.getLanguage("stscript"));
 
+// Enable a script by ID
+const enableScript = (_, scriptId) => {
```
Use `function` blocks like the rest of the code.

p-e-w commented on main.js:
```diff
@@ -201,6 +205,75 @@ function clickHandlerHack() {
+    script.enabled = true;
+    
+    // Update the visual toggle if the script is currently rendered
+    $(`.sorcery-id span:contains('${id}')`).closest('.world_entry').find('.sorcery-enabled')
```
This is not a clean way to find the element. Use a data attribute.

p-e-w commented on main.js:
```diff
@@ -201,6 +205,75 @@ function clickHandlerHack() {
+};
+
+// Register the slash commands
+registerSlashCommand('sorcery-enable', enableScript, [], 'Enable a Sorcery script by ID number', true, true);
```
Wrong quotes.

p-e-w commented on script.html:
```diff
@@ -45,4 +46,4 @@
         </div>
     </div>
-</div>
+</div>
```
:fire:

p-e-w replied:
Fix the EOF please.

p-e-w commented on style.css:
```diff
@@ -60,3 +60,20 @@
+
+.sorcery-id {
+    font-size: 0.85em;
+    background-color: rgba(0, 0, 0, 0.2);
```
Don't hardcode colors, this will look wrong with a different theme.

p-e-w commented on main.js:
```diff
@@ -238,10 +311,14 @@ $(async () => {
         scriptsElement.prepend(scriptElement);
-
+		
+		// Add the script ID to the display
```
Indentation.

p-e-w commented on main.js:
```diff
@@ -238,10 +311,14 @@ $(async () => {
 
-    function addScriptElement(script) {
+    function addScriptElement(script) {	
+    	
```
:fire:

p-e-w commented on main.js:
```diff
@@ -201,6 +205,75 @@ function clickHandlerHack() {
+        toastr.warning("Please provide a script ID to enable");
+        return;
+    }
+    
```
:fire:

p-e-w commented on style.css:
```diff
@@ -60,3 +60,20 @@
+.sorcery-id span {
+    font-weight: bold;
+    color: #ccc;
+}
```
Must end with newline per Unix conventions.

p-e-w reviewed (changes requested):
Please watch out for code style (blank lines, indentation, quotes etc.). It needs to match the rest of the code.

I'm on the fence for what the right command syntax is. Perhaps having a single `sorcery` command with subcommands is better than having `sorcery-enable`, `sorcery-disable` etc.

I would expect the `enable`/`disable` commands to enable/disable Sorcery globally when run without an argument.

bolmin70 committed:
Fixes

Fixed the indentations, added theme colors instead of hardcoded ones, implemented data attributes and used function blocks.

bolmin70 commented:
I fixed and implemented everything that you outlined. As for the functionality of the commands I decided to make a separate enable and disable command to be inline with the /qr-set-on and /qr-set-off commands, which work on a similar basis and are part of the core package. Now I realise that maybe better names would be /sorcery-script-on and /sorcery-script-off, let me know what do you think. As for the the command to enable/disable sorcery globally, in my opinion it should be separate.

bolmin70 committed:
Fixed last empty newline

p-e-w commented on main.js:
```diff
@@ -238,10 +305,15 @@ $(async () => {
     }
 
-    function addScriptElement(script) {
+    function addScriptElement(script) {	   	
```
:fire:

p-e-w commented on script.html:
```diff
@@ -5,6 +5,7 @@
                 <div class="inline-drawer-toggle fa-fw fa-solid fa-circle-chevron-down inline-drawer-icon down">
                 </div>
                 <div class="sorcery-enabled fa-solid killSwitch" title="Enable/disable script"></div>
+                <div class="sorcery-id" title="Script ID" data-script-id="">#<span></span></div>
```
No need to initialize the `data-script-id` attribute here.

p-e-w commented on style.css:
```diff
@@ -60,3 +60,20 @@
+    margin-right: 8px;
+    display: inline-flex;
+    align-items: center;
+    justify-content: center;
```
What is the purpose of using flexbox here? What happens without it?

p-e-w commented on style.css:
```diff
@@ -60,3 +60,20 @@
+
+.sorcery-id span {
+    font-weight: bold;
+    color: #ccc;
```
No hardcoded colors.

p-e-w reviewed (changes requested):
> I decided to make a separate enable and disable command to be inline with the /qr-set-on and /qr-set-off commands, which work on a similar basis and are part of the core package.

Ok, I didn't realize that's an established convention. This is fine then. I prefer the name `sorcery-enable` to `sorcery-script-on`, because it's shorter and matches the language in the UI.

> As for the the command to enable/disable sorcery globally, in my opinion it should be separate.

Is that also a convention established by core? Because that doesn't really make sense to me. What's the upside of having separate commands?

p-e-w commented on main.js:
```diff
@@ -201,6 +202,72 @@ function clickHandlerHack() {
+    saveSettingsDebounced();
+};
+
+// Register the slash commands
```
Comments should explain what the code does, not repeat the code. "Register the slash commands" followed by `registerSlashCommand` is just noise.

p-e-w commented on main.js:
```diff
@@ -238,10 +305,15 @@ $(async () => {
         scriptsElement.prepend(scriptElement);
-
+		
+        // Add the script ID to the data attribute
```
See above regarding comments.

bolmin70 committed:
Created helper function and addressed other issues

Created a helper function for handling changing the script states. Removed unnecassary comments. Removed the data attribute in html file. Fixed styles.

bolmin70 commented:
Addressed everything you outlined. As for the separate command to change global state of sorcery, I just thought it'd make the syntax more user-friendly, but you are right there are no conventions or other upsides related to that. Also I've been thinking of adding a parameter to the commands that the qr-set-on/off ones have - "visible=true/false" which decide if the command should generate the "toast" notifications or not. I'm not sure how to add these extra parameters when registering a command, so I'd have to do some research. What do you think?

p-e-w commented on main.js:
```diff
@@ -201,6 +202,49 @@ function clickHandlerHack() {
+   
+    toastr.success(`Script with ID ${id} has been ${enabled ? "enabled" : "disabled"}`);
+    saveSettingsDebounced();
+    return true;
```
What does this return statement do?

bolmin70 replied:
I thought it'd be convenient for later if the function would return false if changing state failed for some reason and true if it succeeds.

p-e-w replied:
That makes sense in principle, but where else would that function be used?

In fact, I anticipate that the whole thing will need to be completely refactored to add support for `/sorcery-run` in order to avoid duplicating all that code again. This is probably not the right abstraction ATM.

bolmin70 replied:
Yeah you are right about refactoring it to accomodate sorcery-run. If I'll have time to add that command I'll look into it.

p-e-w replied:
You probably want to change `changeScriptState` to `getScript` or something (don't like that name much because it's similar to `addScript` but has different semantics), which returns a script object that you can then manipulate and takes care of the validation and notifications.

bolmin70 replied:
I've thought about it and I'm not sure now what would be the use case of having a sorcery-run command? If someone wants to run a QR there is the /run command and for testing if the sorcery script works there is the already added "Run" button. What purpose would it serve realistically?

p-e-w replied:
It would turn Sorcery into a more powerful version of QR that can also execute JavaScript.

p-e-w replied:
(Ironically, I don't see the purpose of the `/sorcery-enable`/`disable` commands, especially since you have to look up the ID before using them, at which point you can just click the toggle.)

bolmin70 replied:
I explained my use case a bit in our chat, but if you're curious you could check out Len Anderson's Codex extension (interactive maps) as thats where it makes sense to automatically enable/disable scripts based on, for example the current player location, as well as many other factors users can create using it. 

In terms of turning sorcery into more powerful QR's I dont think it should be within the scope of this project (only my opinion of course), as in base QRs have way different functionality than sorcery (can be auto-executed on startup, turned into buttons, etc). I think Sorcery could be implemented directly into the QR editor, but I dont know how inline it is with your idea for the extension. As an example, currently I create a QR (so that it's in the ecosystem, so to speak) and then in Sorcery I just use /run "myQR" |

p-e-w commented on style.css:
```diff
@@ -60,3 +60,16 @@
     grid-area: 1/1;
 }
+
+.sorcery-id {
```
Scope these within `#sorcery-button` like the other style rules in order to prevent polluting the rest of the application.

p-e-w reviewed (changes requested):
> As for the separate command to change global state of sorcery, I just thought it'd make the syntax more user-friendly, but you are right there are no conventions or other upsides related to that.

I think it's very user-friendly to have

```
/sorcery-disable
```

to disable Sorcery, and

```
/sorcery-disable 5
```

to disable the script with ID 5.

> Also I've been thinking of adding a parameter to the commands that the qr-set-on/off ones have - "visible=true/false" which decide if the command should generate the "toast" notifications or not.

Yes, that sounds necessary. It should be possible to run scripts silently. The parameter name `visible` is really bad for this, but if other parts of ST use that it's probably best to use it anyway.

Also, what about `/sorcery-run [scriptId]`?

bolmin70 committed:
Fixed EOF and properly scoped styles

p-e-w commented on script.html:
```diff
@@ -5,6 +5,7 @@
                 <div class="inline-drawer-toggle fa-fw fa-solid fa-circle-chevron-down inline-drawer-icon down">
                 </div>
                 <div class="sorcery-enabled fa-solid killSwitch" title="Enable/disable script"></div>
+                <div class="sorcery-id" title="Script ID">#<span></span></div>
```
This element leads to inconsistent positioning of successive condition inputs when the number of digits in the ID changes (such as between 9 and 10). That looks really bad. Do you have any ideas for fixing this?

bolmin70 replied:
I'm not sure what do you mean? Like when the user removes a script?

p-e-w replied:
When there are more than 10 scripts, the left sides of the condition inputs for scripts `#9` and `#10` will be misaligned, because the width of the label differs.

bolmin70 replied:
Oh I see, I'll look into that. Maybe changing the id system from 1, 2, 3, etc to 001, 002, 003?

bolmin70 replied:
Ok I'll test it. With naming maybe it could follow how world info entries are marked: "UID: X"

p-e-w commented on style.css:
```diff
@@ -60,3 +60,16 @@
+    padding: 2px 6px;
+    border-radius: 8px;
+    background-color: var(--SmartThemeBorderColor);
+    color: var(--SmartThemeEmColor)
```
Missing semicolon.

bolmin70 committed:
Missing semicolon

bolmin70 committed:
Fixes misaligment for double digits

p-e-w commented:
Are you planning to make any more changes here, or should I just merge this and build on top of it?

bolmin70 commented:
Nope, for now I'm done with it, ready to merge.

p-e-w commented:
Ok, I found some time to work on Sorcery today, so I'm merging this. Thanks!

p-e-w commented:
@bolmin70 

I had to revert this PR as simply loading the changes into VSCode revealed several problems:

* The calls to `registerSlashCommand` don't match the signature. No idea what the last two parameters are intended to do.
* `registerSlashCommand` is deprecated. This is documented in https://docs.sillytavern.app/for-contributors/writing-extensions/#registering-slash-commands-new-way
* `toastr` isn't imported and no import appears to be available.

Sorry for this, but I don't have time to untangle all of that, and I can't accept code that litters my IDE with errors. There were also a ton of remaining whitespace issues, even after all the ones I had already pointed out.

bolmin70 commented:
Sure I understand. Probably got outdated through sillytavern updates or something, no idea. I'll just keep using my private version of Sorcery with my changes. Either way, still thanks for everything!
